feat: file watcher for cosmovisor (#8590)
cosmos/cosmos-sdk: cosmovisor/README.md

* `DAEMON_NAME` is the name of the binary itself (e.g. `gaiad`, `regend`, `simd`, etc.).
* `DAEMON_ALLOW_DOWNLOAD_BINARIES` (*optional*), if set to `true`, will enable auto-downloading of new binaries (for security reasons, this is intended for full nodes rather than validators). By default, `cosmovisor` will not auto-download new binaries.
* `DAEMON_RESTART_AFTER_UPGRADE` (*optional*), if set to `true`, will restart the subprocess with the same command line arguments and flags (but with the new binary) after a successful upgrade. By default, `cosmovisor` stops running after an upgrade and requires the system administrator to manually restart it. Note that `cosmovisor` will not auto-restart the subprocess if there was an error.
* `DAEMON_POLL_INTERVAL` is the interval length in milliseconds for polling upgrade plan file. Default: 300.

> * `DAEMON_UPGRADE_INFO_FILE` is a full path to the upgrade plan file created by an app. On start, it checks if the base directory exists. If the file name is wrong the update request will never be handled. Default: `<DAEMON_HOME>/data/upgrade-info.json`.

Reviewer: I'm also a bit confused why someone would want to manually set this and not use the default that is automatically created by the upgrade module. Maybe this could be made more clear in the description here.

> * `DAEMON_UPGRADE_INFO_FILE` is a full path to the upgrade plan file created by an app. On start, it checks if the base directory exists. If the file name is wrong the update request will never be handled. Default: `<DAEMON_HOME>/data/upgrade-info.json`.

Author: We also parse the file. If it has a wrong structure it will error.
I will make it more clear.

> * `DAEMON_UPGRADE_INFO_FILE` is a full path to the upgrade plan file created by an app. On start, it checks if the base directory exists. If the file name is wrong the update request will never be handled. Default: `<DAEMON_HOME>/data/upgrade-info.json`.

Author: > why someone would want to manually set this and not use the default

We can remove it.

> * `DAEMON_UPGRADE_INFO_FILE` is a full path to the upgrade plan file created by an app. On start, it checks if the base directory exists. If the file name is wrong the update request will never be handled. Default: `<DAEMON_HOME>/data/upgrade-info.json`.

Author: We can remove it. In my first design I was using a different file.

│       └── $DAEMON_NAME
└── upgrades
    └── <name>
        ├── bin
        │   └── $DAEMON_NAME

>         └── upgrade_name.txt

Reviewer: ```suggestion
        └── upgrade-info.json
```

The `DAEMON` specific code and operations (e.g. tendermint config, the application db, syncing blocks, etc.) all work as expected. The application binaries' directives such as command-line flags and environment variables also work as expected.

### Detecting Upgrades

> `cosmovisor` is polling the `$DAEMON_HOME/data/upgrade-info.json` file for new upgrade instructions. The file is created by the app x/upgrade module in the `BeginBlocker` when an upgrade is detected and the blockchain reached the upgrade height.

Reviewer: ```suggestion
`cosmovisor` is polling the `$DAEMON_HOME/data/upgrade-info.json` file for new upgrade instructions. The file is created by the x/upgrade module in `BeginBlocker` when an upgrade is detected and the blockchain reaches the upgrade height.
```

The following heuristic is applied to detect the upgrade:

> + When starting, `cosmovisor` doesn't know much about currently running upgrade, except the binary which is `current/bin/` (when it the `current` directory doesn't exists then it is created by linking to `genesis`). It tries to read the `current/update-info.json` file to get information about the current upgrade name.

Reviewer: ```suggestion
+ When starting, `cosmovisor` tries to read the `current/update-info.json` file to get information about the current upgrade name.
```

Author: why do you want to remove the sentence?

Reviewer: It's ok to leave as is, just a suggestion. How the current directory works was explained already a few paragraphs above, so it seemed a bit verbose. It is also a bit confusing referencing "the current upgrade" and then immediately after mentioning the genesis binary (not an upgrade). The essential information is how cosmovisor gets the upgrade name.

Author: I've updated this bullet point.

> + If neither `cosmovisor/current/upgrade-info.json` nor `data/upgrade-info.json` exist then `cosmovisor` will wait for `data/upgrade-info.json` file to trigger an upgrade.

Reviewer: ```suggestion
+ If neither `cosmovisor/current/upgrade-info.json` nor `data/upgrade-info.json` exist, then `cosmovisor` will wait for `data/upgrade-info.json` file to trigger an upgrade.
```

> + If `cosmovisor/current/upgrade-info.json` doesn't exist but `data/upgrade-info.json` exists then `cosmovisor` assumes that whatever is in `data/upgrade-info.json` is a valid upgrade request. In this case `cosmovisor` tries immediately to make an upgrade according to the `name` attribute in `data/upgrade-info.json`.

Reviewer: ```suggestion
+ If `cosmovisor/current/upgrade-info.json` doesn't exist but `data/upgrade-info.json` exists, then `cosmovisor` assumes that whatever is in `data/upgrade-info.json` is a valid upgrade request. In this case `cosmovisor` tries immediately to make an upgrade according to the `name` attribute in `data/upgrade-info.json`.
```

> + Otherwise, we wait for the changes in `upgrade-info.json` - as soon as a new upgrade name will be recorded in that file, we trigger an upgrade mechanism.

Reviewer: ```suggestion
+ Otherwise, `cosmovisor` waits for changes in `upgrade-info.json`. As soon as a new upgrade name is recorded in the file, `cosmovisor` will trigger an upgrade mechanism.
```

> During the upgrade, we auto-download a new binary (if `DAEMON_ALLOW_DOWNLOAD_BINARIES` is enabled) into `cosmovisor/<name>` (where `<name>` is `upgrade-info.json:name` attribute), and link to the `current`. At the end we save `data/upgrade-info.json` to `cosmovisor/current/upgrade-info.json`.

Reviewer: ```suggestion
When the upgrade mechanism is triggered, `cosmovisor` will start by auto-downloading a new binary (if `DAEMON_ALLOW_DOWNLOAD_BINARIES` is enabled) into `cosmovisor/<name>/bin` (where `<name>` is the `upgrade-info.json:name` attribute). `cosmovisor` will then update the `current` symbolic link to point to the new directory and save `data/upgrade-info.json` to `cosmovisor/current/upgrade-info.json`.
```

The `DAEMON` specific code and operations (e.g. tendermint config, the application db, syncing blocks, etc.) all work as expected. The application binaries' directives such as command-line flags and environment variables also work as expected.

### Detecting Updates

`cosmovisor` is pooling the `$DAEMON_UPDATE_INFO_FILE` (default to `$DAEMON_HOME/data/upgrade-info.json`) file for new update instructions. The following heuristic is applied to detect update:
+ When, starting `cosmovisor` doesn't know much about currently running upgrade, except the binary (which is either in `current/bin/` or `genesis/bin` if the former doesn't exists). It tries to read the `current/upgrade_name.txt` file to get an information about the current update name.

> + If `current/upgrade_name.txt` doesn't exist then we assume that `upgrade-info.json` is an upgrade request. So, if the `upgrade-info.json` when starting a cosmovisor but `current/upgrade_name.txt` doesn't exist then we try to make an upgrade according to the `name` attribute in `upgrade-info.json`.

Reviewer: As far as I understand, the use of `upgrade_name.txt` was replaced with `upgrade-info.json`. I think your latest commit is not reflected in the documentation. Can you help clarify this?

Reviewer: ```suggestion
+ If `cosmovisor/current/upgrade-info.json` doesn't exist then `cosmovisor` assumes that `data/upgrade-info.json` is an upgrade request. If `data/upgrade-info.json` when starting a cosmovisor but `cosmovisor/current/upgrade-info.json` doesn't exist, then `cosmovisor` tries to make an upgrade according to the `name` attribute in `data/upgrade-info.json`.
```

+ Otherwise, we wait for the changes in `upgrade-info.json` - as soon as a new upgrade name will be recorded in that file, we trigger an upgrade mechanism.

> During the upgrade we auto-download a new binary (if auto-download is enabled), and link a new directory to the `current` based on the `upgrade-info.json:name`. At the end we save `upgrade-info.json:name` to `current/upgrade_name.txt`.

Reviewer: Again, I think the changes in the latest commit are not reflected here.
```suggestion
During the upgrade, we auto-download a new binary (if auto-download is enabled), and link a new directory to the `current` symbolic link based on the `upgrade-info.json:name`. At the end we save `data/upgrade-info.json` to `cosmovisor/current/upgrade-info.json`.
```
